### PR TITLE
Android keyboard support

### PR DESCRIPTION
**Pull request by ili**
## What does the pull request do?
Implements basic Android soft keyboard support with different input types.

## What is the current behavior?
Soft keyboard is not supported

## What is the updated/expected behavior with this PR?
Support Android keyboard (not IME!) 

## How was the solution implemented (if it's not obvious)?

* `Avalonia.Input` implementation

### CODE REVIEW

**Comment by ili** [comment]
@kekekeks  please review this PR

I do understand that it is far away from the endpoint, but keyboard support is needed, and it could be temporary solution

Next steps may be to expand IME API to support all features, but it is a point for a long term discussion, while even very simple solution is needed to use Avalonia.

**Comment by Gillibald** [comment]
What prevents you from using existing API? https://github.com/AvaloniaUI/Avalonia/tree/master/src/Avalonia.Input/TextInput

In my opinion this PR would introduce unnecessary fragmentation.

**Comment by kekekeks** [comment]
@ili you need to implement `ITopLevelImplWithTextInputMethod` on the Android toplevel implementation. The rest will be managed by our input framework.

Also note that while `TextBeforeCursor` and  `TextAfterCursor` from `ITextInputMethodClient` are not yet implemented by our TextBox, they are intended for Android's ` getTextBeforeCursor(int,int)` and ` getTextAfterCursor(int, int)` IME callbacks .

**Comment by ili** [comment]
@kekekeks sorry for silly question, but i can't find how does `TextInputContentType` is set for `TextBox`

And speaking about `ITextInputMethodImpl` it is not clear how to commit (edit) text using this contract, i see X11 inplemetations has `public event Action<string> Commit`, but it is forwarded to `RawTextInputEvent`, `RawTextInputEvent` should be treated as only edit text API?

**Comment by kekekeks** [comment]
Everything not-yet-commited is considered to be a "preedit" text and should be communicated to `ITextInputMethodClient::SetPreeditText` if it supports preedit text. Once text input session is commited, `RawTextInputEvent` should be triggered.

**Comment by ili** [comment]
@kekekeks thanks!

And what about my silly question, about how can i set TextInputContentType for TextBox? :)

**Comment by kekekeks** [comment]
There is no proper API for setting it from XAML yet, one has to handle `TextInputOptionsQuery` routed event. 

The general shape of the XAML-oriented API would be setting TextInputOptions.XXX attached properties, but that's not implemented yet.

**Comment by ili** [comment]
@kekekeks i see! Where can i see an example how to make it right. 

I'v just added property:

```cs
        public static readonly DirectProperty<TextBox, InputType> InputTypeProperty =
            AvaloniaProperty.RegisterDirect<TextBox, InputType>(
                    nameof(InputType),
                    o => o.InputType);
```

But it looks like not the thing you are talking about, also is it possible to attah this property to `TextBox` and inherit somehow in `Autocomplete`?

**Comment by ili** [comment]
@kekekeks @Gillibald Avalonia.Input implemented, please review

File: src/Android/Avalonia.Android/AndroidInputMethod.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                {
+                    outAttrs.InitialCapsMode = global::Android.Text.CapitalizationMode.Sentences;
+                    outAttrs.InputType |= global::Android.Text.InputTypes.TextFlagCapSentences;
+                }
+
+                if (options.Multiline)
+                    outAttrs.InputType |= global::Android.Text.InputTypes.TextFlagMultiLine;
+            });
+
+            _inputElement.PointerReleased += RestoreSoftKeyboard;
```
Comment (kekekeks) [inline src/Android/Avalonia.Android/AndroidInputMethod.cs:87]: Why do you need to subscribe to PointerReleased? Input method management automatically tells you when new element needs input focus.

File: src/Android/Avalonia.Android/AvaloniaActivity.cs [hunk 2]
```diff
@@ -33,7 +33,7 @@ public object Content
             }
         }
 
-        public override bool DispatchKeyEvent(KeyEvent e) =>
-            View.DispatchKeyEvent(e) ? true : base.DispatchKeyEvent(e);
+        //public override bool DispatchKeyEvent(KeyEvent e) =>
```
Comment (kekekeks) [inline src/Android/Avalonia.Android/AvaloniaActivity.cs:36]: Don't comment code. If it's not needed anymore, just remove it.

File: src/Android/Avalonia.Android/Platform/SkiaPlatform/AndroidFramebuffer.cs [hunk 3]
```diff
@@ -32,7 +32,7 @@ public AndroidFramebuffer(Surface surface, double scaling)
             RowBytes = buffer.stride * (Format == PixelFormat.Rgb565 ? 2 : 4);
             Address = buffer.bits;
 
-            Dpi = scaling * new Vector(96, 96);
```
Comment (kekekeks) [inline src/Android/Avalonia.Android/Platform/SkiaPlatform/AndroidFramebuffer.cs:35]: Why is this change required?

File: src/Android/Avalonia.Android/Platform/SkiaPlatform/InvalidationAwareSurfaceView.cs [hunk 4]
```diff
@@ -13,7 +13,6 @@ public abstract class InvalidationAwareSurfaceView : SurfaceView, ISurfaceHolder
         bool _invalidateQueued;
         readonly object _lock = new object();
         private readonly Handler _handler;
-
```
Comment (kekekeks) [inline src/Android/Avalonia.Android/Platform/SkiaPlatform/InvalidationAwareSurfaceView.cs:16]: nit

File: src/Android/Avalonia.Android/Platform/SkiaPlatform/TopLevelImpl.cs [hunk 5]
```diff
@@ -191,6 +182,29 @@ void ISurfaceHolderCallback.SurfaceChanged(ISurfaceHolder holder, Format format,
...
+            private Action<EditorInfo> _initEditorInfo;
+
+            public void InitEditorInfo(Action<EditorInfo> init)
+            {
+                _initEditorInfo = init;
+            }
+
+            public sealed override IInputConnection OnCreateInputConnection(EditorInfo outAttrs)
+            {
+                if (_initEditorInfo == null)
```
Comment (kekekeks) [inline src/Android/Avalonia.Android/Platform/SkiaPlatform/TopLevelImpl.cs:200]: This method is called by Android when it wishes to. Throwing an exception here doesn't really make sense.

File: src/Android/Avalonia.Android/AndroidInputMethod.cs (hunk 1 above)
Comment (ili) [reply src/Android/Avalonia.Android/AndroidInputMethod.cs:87]: When you press Back button Android hides keyboard, to show it again you do need tap input field (this is native Android behaviour). `ITextInputMethodImpl.SetActive()` is not called in this case (focus not changed). That's why I do nees subscribtion. `PointerReleased` is used because `PointerPressed` is handled already and not bypassed.

File: src/Android/Avalonia.Android/AvaloniaActivity.cs (hunk 2 above)
Comment (ili) [reply src/Android/Avalonia.Android/AvaloniaActivity.cs:36]: No problem, I'v commented it to make the point for reviewers to loot it twice :)

File: src/Android/Avalonia.Android/Platform/SkiaPlatform/AndroidFramebuffer.cs (hunk 3 above)
Comment (ili) [reply src/Android/Avalonia.Android/Platform/SkiaPlatform/AndroidFramebuffer.cs:35]: Because i do receive compiler error. It looks like `scaling * new Vector(96, 96)` returns `double` not `Vector` & `new Vector(96, 96) * scaling`.

File: src/Android/Avalonia.Android/Platform/SkiaPlatform/InvalidationAwareSurfaceView.cs (hunk 4 above)
Comment (ili) [reply src/Android/Avalonia.Android/Platform/SkiaPlatform/InvalidationAwareSurfaceView.cs:16]: Unclear... Do you mean two empty lines needed after fields declaration? :)

File: src/Android/Avalonia.Android/Platform/SkiaPlatform/TopLevelImpl.cs (hunk 5 above)
Comment (ili) [reply src/Android/Avalonia.Android/Platform/SkiaPlatform/TopLevelImpl.cs:200]: This metod is called to create `InputConnection` with IME. Exception is needed because we should configure connection from one side from one side, and as a part of documentation from another: `IInitEditorInfo` is not really clear solution, but i do have no idea how to delegate this better...

May be implementing ITextInputMethodImpl by `ViewImpl`?

Comment (ili) [reply src/Android/Avalonia.Android/Platform/SkiaPlatform/TopLevelImpl.cs]: Yap, you were right... in AndroidTestApplication this method was called during startup...

File: src/Android/Avalonia.Android/AndroidInputMethod.cs (hunk 1 above)
Comment (kekekeks) [reply src/Android/Avalonia.Android/AndroidInputMethod.cs:87]: How does the normal text view behave in that case? Does it keep the focus?

Comment (kekekeks) [reply src/Android/Avalonia.Android/AndroidInputMethod.cs:87]: Note, that we can't just assume which element is the actual text view

Comment (ili) [reply src/Android/Avalonia.Android/AndroidInputMethod.cs:87]: > How does the normal text view behave in that case? Does it keep the focus?

can't say for now. I'll exam it with the real keyboard today.

Comment (ili) [reply src/Android/Avalonia.Android/AndroidInputMethod.cs:87]: @kekekeks i'v checked few android applications, and in some input looks inactive when keyboard is hidden in other - disabled. So it is not clear if it accepts input.

Also soft keyboard is not shown at all when real keyboard is connected.

P.S. FUI tabulation works fine with real keyboard :)

Comment (kekekeks) [reply src/Android/Avalonia.Android/AndroidInputMethod.cs:87]: So we should probably just reset input focus when soft keyboard is dismissed

Comment (ili) [reply src/Android/Avalonia.Android/AndroidInputMethod.cs:87]: I think it can be done on back button pressed handling, but this should be done in Avallnia's input framework

Comment (ili) [reply src/Android/Avalonia.Android/AndroidInputMethod.cs]: @kekekeks  please review the latest changes (`SoftKeyboardListner`). I'll remove event subscribtion related code if this new approach would be consistent.

**Review by kekekeks — APPROVED** [review]
